src/storage/s3.rs

```diff
@@ -76,44 +86,53 @@ pub(super) struct S3StorageTransaction<'a> {
-
-        loop {
-            let mut futures = FuturesUnordered::new();
-            for blob in batch {
-                futures.push(
-                    self.s3
-                        .client
-                        .put_object(PutObjectRequest {
-                            bucket: self.s3.bucket.to_string(),
-                            key: blob.path.clone(),
-                            body: Some(blob.content.clone().into()),
-                            content_type: Some(blob.mime.clone()),
-                            content_encoding: blob.compression.as_ref().map(|alg| alg.to_string()),
-                            ..Default::default()
-                        })
-                        .inspect(|_| {
-                            crate::web::metrics::UPLOADED_FILES_TOTAL.inc_by(1);
-                        }),
-                );
-            }
-            attempts += 1;
-
-            match S3_RUNTIME.lock().block_on(futures.map(drop).collect()) {
-                // this batch was successful, start another batch if there are still more files
-                Ok(_) => break,
-                Err(err) => {
-                    error!("failed to upload to s3: {:?}", err);
-                    // if a futures error occurs, retry the batch
-                    if attempts > 2 {
-                        panic!("failed to upload 3 times, exiting");
+    fn store_batch(&mut self, mut batch: Vec<Blob>) -> Result<(), Error> {
+        self.s3.runtime.handle().block_on(async {
+            // Attempt to upload the batch 3 times
+            for _ in 0..3 {
+                let mut futures = FuturesUnordered::new();
+                for blob in batch.drain(..) {
+                    futures.push(
+                        self.s3
+                            .client
+                            .put_object(PutObjectRequest {
+                                bucket: self.s3.bucket.to_string(),
+                                key: blob.path.clone(),
+                                body: Some(blob.content.clone().into()),
+                                content_type: Some(blob.mime.clone()),
+                                content_encoding: blob
+                                    .compression
+                                    .as_ref()
+                                    .map(|alg| alg.to_string()),
+                                ..Default::default()
+                            })
+                            .map_ok(|_ok| {
+                                crate::web::metrics::UPLOADED_FILES_TOTAL.inc_by(1);
+                            })
+                            .map_err(|err| {
+                                log::error!("Failed to upload blob to S3: {:?}", err);
+                                // Reintroduce failed blobs for a retry
+                                blob
+                            }),
+                    );
+                }
+
+                while let Some(result) = futures.next().await {
```

Comment:
Same comment, this waits on each in serial. Can we wait in parallel instead? That was the whole point of switching to async in the first place.

Reply:
This also doesn't wait in serial since the containing collection is a [`FuturesUnordered`](https://docs.rs/futures-util/0.3.5/futures_util/stream/struct.FuturesUnordered.html). Calling `.poll_next()` (and transitively `.next()`) polls all contained futures until a ready one is found

Reply:
`FuturesUnordered` waits on all futures in itself concurrently.

Reply:
I'm still not quite following ... but @Nemo157 is I think one of the authors of `futures` so I'll take his word for it 😆

Reply:
For the record, Nemo explained it in more depth here: https://discordapp.com/channels/442252698964721669/541978667522195476/736708951332225226